### PR TITLE
Email Service Refactor

### PR DESCRIPTION
**Pull request by MichaelChung123**
# Description
Consolidating all of the different reasons for sending an email from CHEFS into a single function. Also changed the Title for status changed emails to include the form name.

https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-2007

## Types of changes
Refactored code introduces more efficient ways of handling emails being sent out from CHEFS. It also allows for developers to more simply add their own email types by having an interchangeable body content section in their emails while retaining the same wrapper template for the emails.

## Checklist

- [X] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by bcgov-nr-csst** [comment]
PR Deployed at: https://chefs-dev.apps.silver.devops.gov.bc.ca/pr-181

File: app/src/forms/email/emailService.js [hunk 1]
```diff
@@ -3,176 +3,110 @@ const fs = require('fs');
...
+      context: {
+        confirmationNumber: submission.confirmationId,
+        title: configData.title,
+        messageLinkText: configData.messageLinkText,
+        messageLinkUrl: `${service._appUrl(referer)}/form/view?s=${submission.id}`,
+      },
+      to: contextToVal
+    }
+  ];
+  return contexts;
```
Comment (jujaga) [inline app/src/forms/email/emailService.js:29]: As this constant is only used once, you can shorthand this by directly returning the array instead (saves a bit of garbage cleanup time):
```js
return [{
  context: {
    confirmationNumber: submission.confirmationId,
    title: configData.title,
    messageLinkText: configData.messageLinkText,
    messageLinkUrl: `${service._appUrl(referer)}/form/view?s=${submission.id}`,
  },
  to: contextToVal
}];
```

File: app/tests/unit/forms/email/emailService.spec.js [hunk 2]
```diff
@@ -1,116 +1,129 @@
...
 
-  it('should not call chesService if there is a blank array', async () => {
-    form.submissionReceivedEmails = [];
-    chesService.merge = jest.fn().mockReturnValue('sent');
-
-    await emailService._sendSubmissionReceived(form, submission, referer);
-    expect(chesService.merge).toHaveBeenCalledTimes(0);
+    expect(result).toBeTruthy();
+    expect(chesService.merge).toHaveBeenCalledTimes(1);
+    mockedReadFileSync.mockRestore();
```
Comment (jujaga) [inline app/tests/unit/forms/email/emailService.spec.js:110]: You'll want to move `mockedReadFileSync.mockRestore();` into an afterEach section as it applies to all tests within this describe:
```js
afterEach(() => {
   mockedReadFileSync.mockRestore();
});
```

**Review by jujaga — APPROVED** [review]
Looks pretty good codewise. Just a few minor changes to take a look at, and a few indentation issues on the html templates. Tentatively approving this. 👍

File: app/src/forms/email/emailService.js [hunk 3]
```diff
@@ -3,176 +3,110 @@ const fs = require('fs');
...
+  } else if (type === 'sendSubmissionReceived' && configData.form.submissionReceivedEmails) {
+    contextToVal = configData.form.submissionReceivedEmails;
+  }
+  const contexts = [
+    {
+      context: {
+        confirmationNumber: submission.confirmationId,
+        title: configData.title,
+        messageLinkText: configData.messageLinkText,
+        messageLinkUrl: `${service._appUrl(referer)}/form/view?s=${submission.id}`,
```
Comment (loneil) [inline app/src/forms/email/emailService.js:24]: We'll have to adjust this so the submission confirmation email sends the user back to /success?

**Review by loneil — CHANGES_REQUESTED** [review]
Think we need to adjust a url, see comment

File: app/src/forms/email/emailService.js (hunk 3 above)
Comment (jujaga) [reply app/src/forms/email/emailService.js:24]: Good catch - I completely missed this.

Comment (MichaelChung123) [reply app/src/forms/email/emailService.js]: `


const generateContexts = (type, configData, submission, referer) =&gt; {  let contextToVal = [];  let userTypePath = '';  if (type === 'sendStatusAssigned') {    contextToVal = [configData.assignmentNotificationEmail];  } else if (type === 'sendSubmissionConfirmation' &amp;&amp; configData.form.showSubmissionConfirmation) {    contextToVal = [configData.body.to];    userTypePath = '/view?s=';  } else if (type === 'sendSubmissionReceived' &amp;&amp; configData.form.submissionReceivedEmails) {    contextToVal = configData.form.submissionReceivedEmails;    userTypePath = '/success?s=';  }  return [{    context: {      confirmationNumber: submission.confirmationId,      title: configData.title,      messageLinkText: configData.messageLinkText,      messageLinkUrl: `${service._appUrl(referer)}/form${userTypePath}${submission.id}`,    },    to: contextToVal  }];};

`

I added some logic to the generateContexts function in the emailService.js to append /view or /success to those different situtations.

File: app/src/forms/email/emailService.js [hunk 4]
```diff
@@ -2,177 +2,111 @@ const chesService = require('../../components/chesService');
...
-const service = {
+// Helper function used to generate contexts for the ches.merge data argument
+const generateContexts = (type, configData, submission, referer) => {
+  let contextToVal = [];
+  let userTypePath = '';
+  if (type === 'sendStatusAssigned') {
+    contextToVal = [configData.assignmentNotificationEmail];
+  } else if (type === 'sendSubmissionConfirmation' && configData.form.showSubmissionConfirmation) {
+    contextToVal = [configData.body.to];
+    userTypePath = '/view?s=';
```
Comment (loneil) [inline app/src/forms/email/emailService.js:16]: This should be /success

File: app/src/forms/email/emailService.js [hunk 5]
```diff
@@ -2,177 +2,111 @@ const chesService = require('../../components/chesService');
...
+  } else if (type === 'sendSubmissionReceived' && configData.form.submissionReceivedEmails) {
+    contextToVal = configData.form.submissionReceivedEmails;
+    userTypePath = '/success?s=';
```
Comment (loneil) [inline app/src/forms/email/emailService.js:19]: this should be /view

File: app/src/forms/email/emailService.js (hunk 4 above)
Comment (MichaelChung123) [reply app/src/forms/email/emailService.js]: made the swap

File: app/src/forms/email/emailService.js (hunk 5 above)
Comment (MichaelChung123) [reply app/src/forms/email/emailService.js]: made the swap

**Review by TimCsaky — APPROVED** [review]
looks good

File: app/src/forms/email/emailService.js [hunk 6]
```diff
@@ -2,177 +2,111 @@ const chesService = require('../../components/chesService');
...
+  } else if (type === 'sendSubmissionReceived' && configData.form.submissionReceivedEmails) {
+    contextToVal = configData.form.submissionReceivedEmails;
+    userTypePath = '/view?s=';
+  }
+  return [{
+    context: {
+      confirmationNumber: submission.confirmationId,
+      title: configData.title,
+      messageLinkText: configData.messageLinkText,
+      messageLinkUrl: `${service._appUrl(referer)}/form${userTypePath}${submission.id}`,
```
[the remark this answers is not in the document]
Comment (MichaelChung123) [reply app/src/forms/email/emailService.js:27]: added /view to status changed

**Review by loneil — APPROVED** [review]
Looks good, will be very helpful to have the body templates separated out for future work.

**Pull request merged**